Lesson_Materials/Lecture_04_ND_Range_Kernel/index.html

```diff
@@ -317,87 +318,156 @@
 				<section>
 					<div class="hbox" data-markdown>
-						#### Accessing Data With Accessors
+						#### Expressing parallelism
 					</div>
-					<div class="container" data-markdown>
-					* There are a few different ways to access the data represented by an accessor
-					  *  The subscript operator can take an **id**
-					    * Must be the same dimensionality of the accessor
-					    * For dimensions > 1, linear address is calculated in row major 
-					* Nested subscript operators can be called for each dimension taking a **size_t**
-					  * E.g. a 3-dimensional accessor: acc[x][y][z] = …
-					* A pointer to memory can be retrieved by calling **get_pointer**
-					  * This returns a **multi_ptr**, which is a wrapper class for pointers to the memory in the relevant memory space
+					<div class="container">
+						<div class="col">
+							<code><pre>
+auto global_range=sycl::range&lt;2&gt;<mark>(1024, 64)</mark>;
+auto wkgroup_sz=sycl::range&lt;2&gt;<mark>(64, 64)</mark>;
+
+// <mark>OK</mark> - local range divides global range 
+// component wise
+auto my_nd = sycl::ndrange(global_range, wkgroup_sz); 
+
+q.parallel_for&ltkernel&gt(my_nd,[=](<mark>nd_item&lt1&gt ndItem</mark>){
+    /* kernel function code */
+});
+							</code></pre>
+							<code><pre>
+auto global_range=sycl::range&lt;2&gt;<mark>(1000)</mark>;
+auto wkgroup_sz=sycl::range&lt;2&gt;<mark>(64)</mark>;
+
+// <mark>Not OK</mark> - local range doesn't divide divide 
+// global range component-wise
+auto my_nd = sycl::ndrange(global_range, wkgroup_sz); 
+
+q.parallel_for&ltkernel&gt(my_nd,[=](<mark>nd_item&lt1&gt ndItem</mark>){
+    /* kernel function code */
+});
+							</code></pre>
+						</div>
+						<div class="col" data-markdown>
+							* On most hardware, `global_range[i]` must be a multiple of `local_range[i]`.
+							* For Nvidia hardware, workgroup sizes are best chosen from 8, 16, 32, 64, 128, 256, 512, 1024.
+						</div>
+					</div>
+				</section>
+				<!--Slide 20-->
+				<section>
+					<div class="hbox" data-markdown>
+						#### Using Local Memory
+					</div>
+					<div class="container">
+						<div class="col">
+							<code><pre>
+
+q.submit([&](sycl::handler &cgh) {
+  
+  // Local memory is declared using accessors 
+  // with the template parameter
+  // sycl::access::target::local
+  sycl::accessor&lt;T, 1, 
+    sycl::access::mode::read_write, 
+    <mark>sycl::access::target::local&gt;</mark>
+    local_mem(sycl::range<1>(local_mem_sz), 
+  	cgh);  
+
+  cgh.parallel_for(my_ndrange, my_depEvs, 
+    my_kernel);
+});
+							
+							</code></pre>
+						</div>
+						<div class="col" data-markdown>
+							* SYCL local memory is called shared memory in CUDA.
+							* To use local memory you must use the buffer/accessor memory model.
+							* Instead of directly enqueueing tasks to the `queue`, you must `submit` tasks, and use the `sycl::handler` to take care of memory accesses, as well as `single_task` or `parallel_for` submissions.
```

Comment:
```suggestion
							* Instead of directly enqueueing tasks to the `queue`, you must `submit` a command group, and use the `sycl::handler` to allocate the local memory, and invoke `single_task` or `parallel_for` from the `sycl::handler`.
```